### PR TITLE
Updated Cheddar make transaction handling less specific to AWS infrastructure

### PR DESCRIPTION
**Pull request by clicktravel-basharat**

### CODE REVIEW

File: cheddar/cheddar-integration-aws/src/main/java/com/clicktravel/cheddar/infrastructure/messaging/tx/TransactionalMessagePublisher.java [hunk 1]
```diff
@@ -59,7 +58,7 @@ public void begin() throws TransactionException {
     public void commit() throws TransactionException {
         final MessagingTransaction transaction = getCurrentTransaction();
         logger.trace("Committing transaction: " + transaction.transactionId());
-        transaction.applyActions(snsMessagePublisher);
+        transaction.applyActions(messagePublisher);
         currentTransaction.remove();
         logger.trace("Transaction successfully commit: " + transaction.transactionId());
```
Comment (clicktravel-steffan) [inline cheddar/cheddar-integration-aws/src/main/java/com/clicktravel/cheddar/infrastructure/messaging/tx/TransactionalMessagePublisher.java]: Replace 'commit' with 'committed'

File: cheddar/cheddar-integration-aws/src/main/java/com/clicktravel/cheddar/infrastructure/messaging/tx/TransactionalMessageSender.java [hunk 2]
```diff
@@ -59,7 +58,7 @@ public void begin() throws TransactionException {
     public void commit() throws TransactionException {
         final MessagingTransaction transaction = getCurrentTransaction();
         logger.trace("Committing transaction: " + transaction.transactionId());
-        transaction.applyActions(sqsMessageSender);
+        transaction.applyActions(messageSender);
         currentTransaction.remove();
         logger.trace("Transaction successfully commit: " + transaction.transactionId());
```
Comment (clicktravel-steffan) [inline cheddar/cheddar-integration-aws/src/main/java/com/clicktravel/cheddar/infrastructure/messaging/tx/TransactionalMessageSender.java]: Replace 'commit' with 'committed'

File: cheddar/cheddar-integration-aws/src/main/java/com/clicktravel/cheddar/infrastructure/persistence/database/tx/TransactionalDatabaseTemplate.java [hunk 3]
```diff
@@ -67,7 +66,7 @@ public void commit() throws TransactionException {
         final DatabaseTransaction transaction = getCurrentTransaction();
         logger.trace("Committing transaction: " + transaction.transactionId());
         try {
-            transaction.applyActions(dynamoDbTemplate);
+            transaction.applyActions(databaseTemplate);
             currentTransaction.remove();
             logger.trace("Transaction successfully commit: " + transaction.transactionId());
```
Comment (clicktravel-steffan) [inline cheddar/cheddar-integration-aws/src/main/java/com/clicktravel/cheddar/infrastructure/persistence/database/tx/TransactionalDatabaseTemplate.java]: Replace 'commit' with 'committed'

File: cheddar/cheddar-integration-aws/src/main/java/com/clicktravel/cheddar/infrastructure/persistence/filestore/tx/TransactionalFileStore.java [hunk 4]
```diff
@@ -25,23 +24,22 @@
...
-import com.clicktravel.cheddar.infrastructure.persistence.filestore.InternetFileStore;
+import com.clicktravel.cheddar.infrastructure.persistence.filestore.FileStore;
 import com.clicktravel.cheddar.infrastructure.tx.NestedTransactionException;
 import com.clicktravel.cheddar.infrastructure.tx.NonExistentTransactionException;
 import com.clicktravel.cheddar.infrastructure.tx.TransactionException;
 import com.clicktravel.cheddar.infrastructure.tx.TransactionalResource;
-import com.clicktravel.infrastructure.persistence.aws.s3.S3FileStore;
 
-public class TransactionalS3FileStore implements InternetFileStore, TransactionalResource {
+public class TransactionalFileStore implements FileStore, TransactionalResource {
```
Comment (clicktravel-steffan) [inline cheddar/cheddar-integration-aws/src/main/java/com/clicktravel/cheddar/infrastructure/persistence/filestore/tx/TransactionalFileStore.java]: I see the contract has changed, this class no longer supports InternetFileStore. Is there such a thing as a TransactionalInternetFileStore?

Comment (clicktravel-basharat) [inline cheddar/cheddar-integration-aws/src/main/java/com/clicktravel/cheddar/infrastructure/persistence/filestore/tx/TransactionalFileStore.java]: There is no such thing. We have a similar thing with a BatcheDatabaseTemplates: you cannot make them transactional.

File: cheddar/cheddar-integration-aws/src/main/java/com/clicktravel/cheddar/infrastructure/tx/manager/SimpleTransactionalResourceManager.java [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+
+        // Some constraints in the domain are enforced on 'save' actions e.g. uniqueness constraint
+        // It is unsafe for the remaining commits to proceed if this commit is not completed first
+        commitDatabaseTemplateTransaction();
+
+        commitFileStoreTransaction();
+
+        commitMessageSenderTransaction();
+
+        // By doing this commit last, we ensure domain events are published only if all other actions in this
```
Comment (clicktravel-steffan) [inline cheddar/cheddar-integration-aws/src/main/java/com/clicktravel/cheddar/infrastructure/tx/manager/SimpleTransactionalResourceManager.java]: Remove word 'domain', as this applies to all types of events

File: cheddar/cheddar-integration-aws/src/test/java/com/clicktravel/cheddar/infrastructure/persistence/database/tx/TransactionalDatabaseTemplateTest.java [hunk 6]
```diff
@@ -41,25 +41,23 @@
...
 import com.clicktravel.cheddar.infrastructure.tx.TransactionalResourceException;
-import com.clicktravel.infrastructure.persistence.aws.dynamodb.DynamoDbTemplate;
-import com.clicktravel.infrastructure.persistence.aws.dynamodb.StubItem;
 
 @SuppressWarnings("unchecked")
-public class TransactionalDynamoDbTemplateTest {
+public class TransactionalDatabaseTemplateTest {
 
-    private final DynamoDbTemplate dynamoDbTemplate = mock(DynamoDbTemplate.class);
+    private final DatabaseTemplate dynamoDbTemplate = mock(DatabaseTemplate.class);
```
Comment (clicktravel-steffan) [inline cheddar/cheddar-integration-aws/src/test/java/com/clicktravel/cheddar/infrastructure/persistence/database/tx/TransactionalDatabaseTemplateTest.java]: Rename field to 'databaseTemplate'

File: cheddar/cheddar-integration-aws/src/test/java/com/clicktravel/cheddar/infrastructure/persistence/database/tx/TransactionalDatabaseTemplateTest.java [hunk 7]
```diff
@@ -41,25 +41,23 @@
...
 
     @Test
-    public void shouldCreateTransactionalDynamoDbTemplate_withDynamoDbTemplate() {
+    public void shouldCreateTransactionalDatabaseTemplate_withDatabaseTemplate() {
         // Given
-        final DynamoDbTemplate dynamoDbTemplate = mock(DynamoDbTemplate.class);
+        final DatabaseTemplate dynamoDbTemplate = mock(DatabaseTemplate.class);
```
Comment (clicktravel-steffan) [inline cheddar/cheddar-integration-aws/src/test/java/com/clicktravel/cheddar/infrastructure/persistence/database/tx/TransactionalDatabaseTemplateTest.java]: This local variable appears to be shadowing the field variable

File: cheddar/cheddar-integration-aws/src/test/java/com/clicktravel/cheddar/infrastructure/persistence/database/tx/TransactionalDatabaseTemplateTest.java [hunk 8]
```diff
@@ -546,10 +544,7 @@ private StubItem randomStubItem() {
         final StubItem item = new StubItem();
         item.setId(randomId());
         item.setStringProperty(randomString(10));
-        item.setStringProperty2(randomString(10));
-        item.setBooleanProperty(randomBoolean());
         item.setVersion(randomLong());
-        item.setStringSetProperty(Sets.newSet(randomString(10), randomString(10), randomString(10)));
```
Comment (clicktravel-steffan) [inline cheddar/cheddar-integration-aws/src/test/java/com/clicktravel/cheddar/infrastructure/persistence/database/tx/TransactionalDatabaseTemplateTest.java]: Can you explain why these item.set(...) statements have been removed?

Comment (clicktravel-basharat) [inline cheddar/cheddar-integration-aws/src/test/java/com/clicktravel/cheddar/infrastructure/persistence/database/tx/TransactionalDatabaseTemplateTest.java]: Using a simpler Item bean to illustrate what is happening. It only has one property: stringProperty.

**Comment by clicktravel-steffan** [comment]
@clicktravel-basharat First review completed, see comments above. It might be worth do a case-insensitive search for 'Dynamo' and 'S3' in relevant parts of both main and test source trees to make sure there is no mention of either where they no longer belong.

File: cheddar/cheddar-integration-aws/src/test/java/com/clicktravel/cheddar/infrastructure/persistence/database/tx/TransactionalDatabaseTemplateTest.java (hunk 7 above)
Comment (clicktravel-basharat) [inline cheddar/cheddar-integration-aws/src/test/java/com/clicktravel/cheddar/infrastructure/persistence/database/tx/TransactionalDatabaseTemplateTest.java]: Yes, renamed the instance variable to mockXxx

**Comment by clicktravel-basharat** [comment]
I've also done a search for the AWS technology names and removed them.

**Comment by clicktravel-steffan** [comment]
@clicktravel-basharat Second review complete, just one comment outstanding.

**Pull request merged**